file/writable_file_writer.cc

```diff
@@ -451,6 +514,76 @@ IOStatus WritableFileWriter::WriteBuffered(const char* data, size_t size) {
     src += allowed;
   }
   buf_.Size(0);
+  buffered_data_crc32c_checksum_ = 0;
+  return s;
+}
+
+IOStatus WritableFileWriter::WriteBufferedWithChecksum(const char* data,
+                                                       size_t size) {
+  IOStatus s;
+  assert(!use_direct_io());
+  assert(perform_data_verification_ && buffered_data_with_checksum_);
+  const char* src = data;
+  size_t left = size;
+  DataVerificationInfo v_info;
+  char checksum_buf[sizeof(uint32_t)];
+
+  // Check how much is allowed. Here, we loop until the rate limiter allows to
+  // write the entire buffer.
+  // TODO: need to be improved since it sort of defeats the purpose of the rate
+  // limiter
```

Comment:
The long term rate limit is respected, but the burst limit can be violated with sufficiently large Appends. ("sort of defeats the purpose" is vague.)

If we imagine a scenario with a large Append and want to minimize the damage from writing it all at once, across different scenarios with saturated or unsaturated I/O buffers/queues/etc., I can imagine the most harm coming from  one of these depending on the scenario:
* Execute the write after the full size has been accepted by the rate limiter (current). Under low demand, this could lead to a quiet period followed by a write burst. 
* Execute the write before waiting on the rate limiter. Under high demand, this could excessively burst/queue writes in the I/O system.

Perhaps a compromise implementation would execute the write after at least half of the full size has been accepted by the rate limiter. Just an idea but would add extra implementation complexity.

Reply:
> Perhaps a compromise implementation would execute the write after at least half of the full size has been accepted by the rate limiter. Just an idea but would add extra implementation complexity.

The size of the data being written is hard to decide since it is controlled by when "WriteBuffered" is called. It is possible that, as you suggested, we set a hard limit, when the buffered data size is larger than a certain size, we have the second checksum for the rest of the data. Then, with some identification and control logic, in WriteBuffered, we will call 2 WritableFile::Append with their own checksum (first half and second half). So we will accumulate the credit from rate-limiter twice and relief the burst write in some degree. 

Or, we explicitly have an option to limit the write buffer size when checksum handoff is enabled, so the very large writes can be avoided. What do you think?